New Windows Module: win_certificate_info

mhunsber opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
adding a new windows module `win_cert_stat` that returns information about installed certificates from the windows certificate store.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
win_cert_stat
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
`ansible testhost -m win_cert_stat -a "thumbprint=BD7AF104CF1872BDB518D95C9534EA941665FD27"`
<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
testhost | SUCCESS => {
    "archived": false,
    "changed": false,
    "dns_names": [
        "subject.ansible.com"
    ],
    "exists": true,
    "extensions": [],
    "friendly_name": "",
    "has_private_key": false,
    "issued_by": "root.ansible.com",
    "issued_to": "subject.ansible.com",
    "issuer": "C=US, CN=root.ansible.com",
    "public_key": "MIIBCgKCAQEA7M6nRUqNxpVT9cVJzeG0j6xURYjpfzKrqdGZ+VCSVEfcC50Q/ji3Ku5Cop+oWKqM0ru3AvgZ1kKKu7xkkjVar+wahTMMxibfSAVEL9FKiEJAP3blHSnQIjJ9j0EbDQ55FhG
04+MbIrSLVTb/OHAxtJZGQWMk4H8mZWS2VqZZKVEJK2YaBQnpGUSfjg66jVt1uetK6zlCIfh4kHgaqHGrejncYCAgaqVIfBFaNYoWtw14xjdPKubrYOoJmyJdlQWlSY38ncJ1Eqh8pScIelgiBIfh17J+J3XzVkjf3
w6r4uSNdFdY+maLO8qkH/KgHs1AJunkqoS4Gc9KHavQ7DuBxwIDAQAB",
    "raw": "MIIC0TCCAbkCCQC/MtOBa1UDpzANBgkqhkiG9w0BAQsFADAoMRkwFwYDVQQDDBByb290LmFuc2libGUuY29tMQswCQYDVQQGEwJVUzAgFw0xNzEyMTUwODU2MzBaGA8yMDg2MDEwMjA4NTYzMFowKz
EcMBoGA1UEAwwTc3ViamVjdC5hbnNpYmxlLmNvbTELMAkGA1UEBhMCVVMwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDszqdFSo3GlVP1xUnN4bSPrFRFiOl/Mqup0Zn5UJJUR9wLnRD+OLcq7kKin6hY
qozSu7cC+BnWQoq7vGSSNVqv7BqFMwzGJt9IBUQv0UqIQkA/duUdKdAiMn2PQRsNDnkWEbTj4xsitItVNv84cDG0lkZBYyTgfyZlZLZWplkpUQkrZhoFCekZRJ+ODrqNW3W560rrOUIh+HiQeBqocat6OdxgICBqpU
h8EVo1iha3DXjGN08q5utg6gmbIl2VBaVJjfydwnUSqHylJwh6WCIEh+HXsn4ndfNWSN/fDqvi5I10V1j6Zos7yqQf8qAezUAm6eSqhLgZz0odq9DsO4HHAgMBAAEwDQYJKoZIhvcNAQELBQADggEBAFK5mVIJ2D+k
I0kksxnW4ibWFjzlYFYPYrZg+2JFIVTbKBg1YzyhuIKm0uztqRxQq5iLn/C/uponHoqF7KDQI37KAJIQdgSva+mEuO9bZAXg/eegail2hN6np7HjOKlPu23s40dAbFrbcOWPVbsBEPDP0HLv6OgbQWzNlE9HO1b7pX
6ozk3q4ULO7IR85P6OHYsBBThL+qsOTzg/gVknuB9+n9hgNqZcAcXBLDetOM9aEmYJCGk0enYP5UGLYpseE+rTXFbRuHTPr1o6e8BetiSWS/wcrV4ZF5qr9NiYt5eD6JzTB5Rn5awxxj0FwMtrBu003lLQUWxsuTzz
35/RLY4=",
    "serial_number": "00BF32D3816B5503A7",
    "signature_algorithm": "sha256RSA",
    "subject": "C=US, CN=subject.ansible.com",
    "thumbprint": "BD7AF104CF1872BDB518D95C9534EA941665FD27",
    "valid_from": 1513328190,
    "valid_to": 3660800190,
    "version": 1
}
```

Commit:
win_cert_stat initial commit with tests

mhunsber commented:
A note about the tests. The check for the public key matching the openssl command output is commented out because for some reason the .net `x509certificate2.GetPublicKey()` command was returning fewer bytes than the openssl command, so the comparison was failing.

For both certs I'm getting these extra bytes in the public key from the openssl command (or missing them in the .net object depending on how you look at it
```
34, 48, 13, 6, 9, 42, 134, 72, 134, 247, 13, 1, 1, 1, 5, 0, 3, 130, 1, 15, 0, 48, 130, 1
```
otherwise the byte arrays match, but I don't know enough about certificates to know why this is.

Commit:
documentation fix.
first attempt windows server 2008 compatibility

ShachafGoldstein commented:
Can you check if this might be the issue with the test you mentioned? 
https://stackoverflow.com/questions/54214899/get-public-key-byte-array-from-cer-file-c-sharp

ShachafGoldstein commented on test/integration/targets/win_cert_stat/tasks/tests.yml:
```diff
@@ -0,0 +1,73 @@
+    thumbprint: '{{ subj_thumbprint }}'
+  register: subj_stats
+
+- name: dump stats
```
I guess this is for debug and will be deleted before final?

mhunsber replied:
Correct, that was so I could see the output of the public key and compare it to the openssl output.

ShachafGoldstein commented on lib/ansible/modules/windows/win_cert_stat.py:
```diff
@@ -0,0 +1,225 @@
+  returned: cert exists
+  type: str
+  sample: FF6797793A3CD798DC5B2ABEF56F73EDC9F83A64
+valid_from:
```
Maybe add a parsed value of the dates in addition to the raw one?

mhunsber replied:
Sounds like a good addition. Any suggestions for names?
I think I'd prefer `valid_from_iso8601` or `valid_from_formatted`.
I could also put the formatted date in `valid_from` and the timestamp in `valid_from_unix`.
Or we could make it a complex object.

mhunsber replied:
added valid_from|to_iso8601

ShachafGoldstein commented on lib/ansible/modules/windows/win_cert_stat.ps1:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+$store_location_values = ([System.Security.Cryptography.X509Certificates.StoreLocation]).GetEnumValues() | ForEach-Object { $_.ToString() }
```
More a question than a comment, why did you use `foreach` here but `-as` below?

mhunsber replied:
Probably because I copied it from `win_certificate_store` and never changed the style. It makes sense to have it be consistent though.

mhunsber replied:
I think I found the reason!
https://app.shippable.com/github/ansible/ansible/runs/151798/54/tests

```
 System.ArgumentException: argument spec for 'choices' did not match expected type System.Collections.Generic.List`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]: actual type System.String[] - found in store_location
   at Ansible.Basic.AnsibleModule.ValidateArgumentSpec(IDictionary argumentSpec) in c:\Users\Administrator\AppData\Local\Temp\evhsqs5m.0.cs:line 610
   at Ansible.Basic.AnsibleModule.ValidateArgumentSpec(IDictionary argumentSpec) in c:\Users\Administrator\AppData\Local\Temp\evhsqs5m.0.cs:line 623
   at Ansible.Basic.AnsibleModule..ctor(String[] args, IDictionary argumentSpec) in c:\Users\Administrator\AppData\Local\Temp\evhsqs5m.0.cs:line 195
```

ShachafGoldstein reviewed (commented):
Looks good

mhunsber commented:
I'll can into it, certainly feels related from initial read. The problem description makes me think I should expect the .net object to return the extra bytes, not the other way around. Also, the comments to this question suggest that it might not be worth pursuing.

Commit:
add formatted dates
removed debug tests

Commit:
make choices generic list

mhunsber reviewed (commented):
added the formatted dates and removed the debug dump tests

jborean93 commented on lib/ansible/modules/windows/win_cert_stat.ps1:
```diff
@@ -0,0 +1,103 @@
+
+$cert_path = [io.path]::Combine('Cert:\', $store_location, $store_name, $thumbprint)
+
+if (Test-Path -LiteralPath $cert_path)
```
Using the PowerShell cert provider gives us a bit of a disconnect between the store names used in `win_certificate_store`. It would be nice if we can have both modules aligned.

mhunsber replied:
I agree, and I did originally use the x509store object, but the enum values were too limiting. For example, in a server 2012 instance, I can find 3 certificates in `cert:\localmachine\ca`, but I can't find any of them in the enum values of `AddressBook,AuthRoot,CertificateAuthority,Disallowed,My,Root,TrustedPeople,TrustedPublisher`

However, I _can_ open the 'CA' store using the `x509store` object, even though 'CA' isn't part of the enum.
```
$store = New-Object -TypeName X509Store -ArgumentList 'CA', "LocalMachine"
$store.Open("readOnly")
$store.Certificates # outputs the same as 'ls cert:\localmachine\ca'
```
I wasn't sure how to "test" that a store name existed, or how to generate a list of valid stores. Any ideas?

mhunsber replied:
Okay, turns out you can have [arbitrary store names](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509store.-ctor?view=netframework-2.0#System_Security_Cryptography_X509Certificates_X509Store__ctor_System_String_System_Security_Cryptography_X509Certificates_StoreLocation_), and then create new stores that way as well. For that reason, if the store doesn't exist, trying to open it will result in 0 certificates, so I can switch to using the x509store object and still allow any string for `store_name`. If the store contains 0 certificates, would we prefer the module fail or just return an empty array?

jborean93 replied:
Oh wow, I did not know that. I think we should still document the valid enum stores that are available in the description but don't set it in the choices. Looks like `win_certificate_store` needs to be updated as well.

> If the store contains 0 certificates, would we prefer the module fail or just return an empty array?

I think just return an empty array is fine.

jborean93 commented on lib/ansible/modules/windows/win_cert_stat.ps1:
```diff
@@ -0,0 +1,103 @@
+
+function Format-Date([DateTime]$date)
+{
+    return $date.ToUniversalTime().ToString('yyyy-MM-ddTHH:mm:ssK')
```
You can also do `ToString("o")` which I believe adds the UTC offset.

mhunsber replied:
It does, but it also adds unnecessary precision, and will format it as "local time" + "utc offset", instead of converting it to UTC and noting that it is UTC time.
```
(Get-Date).ToString('o') # 2019-12-04T07:55:28.7763737-08:00
(Get-Date).ToUniversalTime().ToString('yyyy-MM-ddTHH:mm:ssK') # 20119-12-04T15:55:28Z
```
It adds all the partial seconds even when the date-time object has them set to 0
```
(Get-Date -Millisecond 0 -Second 0).tostring('o') # 2019-12-04T11:08:00.0000000-05:00
```
which is why I opted for the custom format string, since most (if not all?) of certificate expiration dates only have a precision of seconds, not milliseconds

jborean93 replied:
Fair enough makes sense

jborean93 commented on lib/ansible/modules/windows/win_cert_stat.py:
```diff
@@ -0,0 +1,235 @@
+
+DOCUMENTATION = r'''
+---
+module: win_cert_stat
```
This should be called `win_certificate_stat` maybe even `win_certificate_info` to better align with the `_info` modules we seem to be headed towards. Doing the `_info` one is preferable as we can include functionality to enumerate all certificates in a store instead of just the 1 a user wants.

jborean93 commented on lib/ansible/modules/windows/win_cert_stat.py:
```diff
@@ -0,0 +1,235 @@
+    choices: [ CurrentUser, LocalMachine ]
+    default: LocalMachine
+seealso:
+- module: win_stat
```
I don't think we need `win_stat` and `win_reg_stat`

jborean93 commented on lib/ansible/modules/windows/win_cert_stat.py:
```diff
@@ -0,0 +1,235 @@
+'''
+
+RETURN = r'''
+archived:
```
If we wanted to go the route of an `_info` module instead we should return a list of certs that contain these entries.

mhunsber replied:
How do I note in the return documentation yml that it is a list type with complex elements?

jborean93 replied:
Have a look at [ec2_instances_info](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_instance_info.py#L75) for how to document a list of complex elements.

jborean93 reviewed (changes requested):
Things look pretty good but I feel that we should instead have this module be `win_certificate_info` where it will enumerate all certs in the store specified. If the thumbprint was specified we would still return a list but it only contains the 1 entry. We can still have the `exists` as a root key to make it easy for people to test if a cert existed or not.

The other thing is I think the store names should align with the .NET based `win_certificate_store`. This might need some slight tweaks to the code here but considering the cert provider for `Get-Item` returns the same type it shouldn't be anything major.

Commit:
return a list of certificates
use .net x509 store instead of PS cert provider

Commit:
fixed tests file

Commit:
fix timestamps returning null

Commit:
rename to win_certificate_info

Commit:
rename tests win_certificate_info

mhunsber commented:
Thanks for the feedback and help, @jborean93. I've renamed the module to win_certificate_info, modified it to use the .NET x509Store object instead of the powershell cert provider to keep in more in line with `win_certificate_store`, and changed it to return a list of certificates.

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+$store = New-Object -TypeName System.Security.Cryptography.X509Certificates.X509Store -ArgumentList $store_name, $store_location
+try
+{
+    $store.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite)
```
Is ReadWrite needed, I would have assumed `Read` is what we need for this exercise.

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+{
+    $store.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite)
+}
+catch [System.Security.Cryptography.CryptographicException]
```
I think I went the wrong route with `win_certificate_store` in trying to catch every exception type that was raised in some of the .NET calls. It was from an era when Ansible's messages for uncaught exceptions were quite minimal and sometimes cryptic.

Since 2.8 we properly format any uncaught exceptions to include things like the stack trace and actual PowerShell message making it easy and simple for people to see:

1. Where the exception was raised,
2. What the exception type was, and
3. The actual line where the exception was raised

This coupled with the fact that we can reduce the lines from 16 to just 1 makes the code a lot simpler. The only times I think a try/catch should be used today is when

1. We can actually recover from the exception,
2. Using it for flow control, to determine the next path, or
3. The exception message returned is quite vague and we need to add some context around it that may not be apparent by default

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+    $cert_info.dns_names = @($cert_info.issued_to)
+    $cert_info.raw = [System.Convert]::ToBase64String($cert.GetRawCertData())
+    $cert_info.public_key = [System.Convert]::ToBase64String($cert.GetPublicKey())
+    $cert_info.extensions = @()
```
Adding to an array can be quite an expensive operation, especially if the array elements are large, or there are lots of elements in the array. You are better of relying on PowerShell's piping mechanism that converts a collection to an array or build the result as an generic list and adding to it.

It's more PowerShell like to do the former and would look something like this

```
$cert_info.extensions = foreach($extension in $cert.Extensions) {
    $extension_info = @{
        critical = $extension.Critical
        field = $extension.Oid.FriendlyName
        value = $extension.Format($false)
    }
    .... # Do the rest of the work you want on the hashtable (if needed)

    # By outputting the hashtable you are adding it to the temp collection which is unpacked to an array
    # on the foreach loop. This means $cert_info.extensions is an array that contains each output you create
    $extension_info
}
```

mhunsber replied:
I did not know that `$result = foreach($elem in $array) { ... }` was valid, and thought you had to do the pipeline `$result = $array | Foreach-Object { ... }`. That does make sense since `$result = if($expr) { ... } else { ... }` and `$result = switch($item) { ... }` are both valid. Thanks!

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+    $cert_info.issued_by = $cert.GetNameInfo('SimpleName', $true)
+    $cert_info.issued_to = $cert.GetNameInfo('SimpleName', $false)
+    $cert_info.signature_algorithm = $cert.SignatureAlgorithm.FriendlyName
+    $cert_info.dns_names = @($cert_info.issued_to)
```
I see you are adding more elements to this array later on, you should never add to an array. Instead use a generic list which is more optimised for operations like this

```suggestion
    $cert_info.dns_names = [System.Collections.Generic.List`1[String]]@($cert_info.issued_to)
```

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+                $san_parts = $san.Split("=")
+                if ($san_parts.Length -ge 2 -and $san_parts[0].Trim() -eq 'DNS Name')
+                {
+                    $cert_info.dns_names += $san_parts[1].Trim()
```
Now that `dns_names` is a list (based on a previous comment), this should be

```suggestion
                    $cert_info.dns_names.Add($san_parts[1].Trim())
```

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.ps1:
```diff
@@ -0,0 +1,145 @@
+    if ($found_certs.Count -gt 0)
+    {
+        $module.Result.exists = $true
+        [array]$module.Result.certificates = $found_certs | ForEach-Object { Get-CertificateInfo -cert $_ }
```
I wonder if it's a good idea to make sure the output is sorted and not rely on whatever algorithm .NET has. If we wanted to it would just be a matter of adding `| Sort-Object -Property {$_.<key>}`

What are your thoughts?

mhunsber replied:
I like the idea of sorting so that it is a predictable order, but I don't know which property makes the most sense to sort by. I think the default in the certificate store GUI is to sort by the "issued to" property.

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.py:
```diff
@@ -0,0 +1,226 @@
+EXAMPLES = r'''
+- name: Obtain information about a particular certificate in the computer's personal store
+  win_certificate_info:
+    thumbprint: BD7AF104CF1872BDB518D95C9534EA941665FD27
```
I'm curious as to whether this needs to be in upper case or if it works in lower case as well.

mhunsber replied:
It can be upper, lower, or mixed case, but it cannot have hyphens separating the bytes, which is how some applications display it. I will note the format

jborean93 commented on lib/ansible/modules/windows/win_certificate_info.py:
```diff
@@ -0,0 +1,226 @@
+    required: no
+  store_name:
+    description:
+    - The name of the store to search.
```
Might want to link to the enum that already have some names they can use as a guide.

jborean93 reviewed (commented):
Looks great, just a few more questions/comments

Commit:
return certificates as a sorted array
open the store with readonly privileges

Commit:
extensions always returned as an array

mhunsber commented:
I've made the requested changes. Failing check seems to be something unrelated (T=azure/2.7/10) ?
@jborean93 I did add the sort, but on thumbprint. I can easily be convinced about a different sort key depending on other people's thoughts.

jborean93 reviewed (approved):
Looks great, thanks for sticking with it

jborean93 commented:
rebuild_merge

jborean93 commented:
ansibot doesn't seem to want to rebuild or merge so I'll just do it myself.
